python/pyarrow/schema.pyx

```diff
@@ -256,34 +299,47 @@ def timestamp(unit_str):
     else:
         raise TypeError('Invalid TimeUnit string')
 
-    if unit in _timestamp_type_cache:
-        return _timestamp_type_cache[unit]
+    cdef TimestampType out = TimestampType()
+
+    if tz is None:
+        out.init(la.timestamp(unit))
+        if unit in _timestamp_type_cache:
```

Comment:
Oh I see, we could have two threads adding `unit` to the cache. Yeah I think it's fine. I was thinking about C-level threadsafety